doc/developers.rst

```diff
@@ -451,6 +444,138 @@ DataAccess class i.e. the `_field_write_arguments()` and
+provided and a particular maximum line length can be specified.
+
+Once an instance of the `Parser()` class is created and configured
+with required values for the optional arguments, then the parse method
+can be called. This takes the name of the input code as its argument
+and returns a parse tree of the input code and a FileInfo object that
+captures the required invoke information found in the input code and
+in the associated kernel codes.
+
+If the `nemo` API is specified then the `parse` method of the `Parser`
+instance simply parses the code with `fparser2` and returns the
+resultant `fparser2` parse tree.
+
+For all other APIs the `parse` method of the `Parser` instance returns
+the resultant `fparser2` parse tree and a `FileInfo` instance which
+captures the invoke and kernel metadata in a hierarchy of classes.
+
+When the `Parser` instance parses the code it expects to find Fortran
+code program, module, subroutine or function (and it aborts if
+not). Currently the first of these (there may be more than one
+subroutine for example) is assumed to be the one that is
+required. This limititation is captured in issue #307.
+
+The native `fparser2` tree of the Fortran code is then walked and all
+use statement names are captured and stored in a map (called
+`_arg_name_to_module_name`). This map allows the module name to be
+found given a particular argument name. Also, if an `invoke` call is
+found then an `InvokeCall` object is created and added to a list of
+such instances. Once all invokes have been found the `FileInfo`
+instance is created.
+
+.. note:: In the future we want to be able to simply replace one parse
+          tree with another. So how should we do this? One option
+          would be to try to minimise the parser-specific parts and
+          create some form of interface. The question is really what
+          level of interface we should use.
+
+An `InvokeCall` object is created in the `create_invoke_call` method
+by first parsing each of the kernels specified in an invoke call and
+creating a list of `kernel` objects which are then used to create
+an `InvokeCall`.
+
+A `kernel` object is created in the `create_kernel_call` method which
+extracts the kernel name and kernel arguments, then creates either a
+`BuiltInCall` instance (via the `create_builtin_kernel_call` method)
+or a `KernelCall` instance (via the `create_coded_kernel_call`
+method). `BuiltInCalls` are created if the kernel name is the same as
+one of those specified in the builtin names for this particular API
+(see the variable `_builtin_name_map` which is initialised by the
+`get_builtin_defs` function.
+
+The `create_kernel_call` method uses the `get_kernel` function to find
+out the kernel name and create a list or `Arg` instances representing
+the arguments. The `get_kernel` function parses each kernel argument
+using the fparser2 AST and determines the required argument
+information. As `fparser2` parses all of a code we can use the parse
```

Comment:
I don't fully understand the last sentence. Are you saying this is something that we could do in future (and thus validate kernel meta-data)?

Reply:
I've updated to (hopefully) clarify.